crates/control/src/localization.rs

```diff
@@ -208,26 +208,44 @@ impl Localization {
                     None => {}
                 };
             }
-            (PrimaryState::Penalized, _, _)
-                if primary_state != PrimaryState::Penalized
-                    && (self.time_when_penalized_clicked.map_or(true, |time| {
+            (PrimaryState::Penalized, _, _) if primary_state != PrimaryState::Penalized => {
+                if self.was_picked_up_while_penalized
+                    || self.time_when_penalized_clicked.map_or(true, |time| {
                         context
                             .cycle_time
                             .start_time
                             .duration_since(time)
                             .expect("time ran backwards")
                             > *context.tentative_penalized_duration
-                    }) || !context.has_ground_contact) =>
-            {
-                if self.is_penalized_with_motion_in_set_or_initial {
-                    if self.was_picked_up_while_penalized_with_motion_in_set_or_initial {
-                        self.hypotheses = take(&mut self.hypotheses_when_entered_playing);
-
+                    })
+                {
+                    if self.is_penalized_with_motion_in_set_or_initial {
+                        if self.was_picked_up_while_penalized {
```

Comment:
Please restructure this so you don't have to check it twice. I think this is quite confusing to read.

Reply:
Thanks, done.

Reply:
Doesn't seem like it, even after switching to the correct branch 😉

Reply:
Could you please elaborate what you mean? The previously included double check of `self.was_picked_up_while_penalized` is not present anymore.